### PR TITLE
feat(*): we now also ship code `es2015` module format for better tree…

### PR DESCRIPTION
**Pull request by alan-agius4**
Ship code in ES2015 module for better treeshaking

Updated tslib  https://github.com/Microsoft/tslib/blob/master/README.md

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -67,9 +69,11 @@
     "typescript": "~2.3.4"
   },
   "peerDependencies": {
-    "lodash": "4.x"
+    "lodash": "4.x",
+    "tslib": "^1.7.1"
   },
   "dependencies": {
-    "tslib": "^1.6.1"
+    "lodash": "^4.17.4",
```
Comment (steelsojka) [inline package.json]: Why add lodash as a dependency and peer?

File: src/after.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import after = require('lodash/after');
```
Comment (steelsojka) [inline src/after.ts:1]: The reason these statements aren't named import statements is to only import the necessary lodash method vs importing all of lodash.

File: src/factory/DecoratorFactory.ts [hunk 3]
```diff
@@ -43,8 +43,8 @@ export class InternalDecoratorFactory {
       return (target: Object, name: string, _descriptor?: PropertyDescriptor): PropertyDescriptor => {
         const descriptor = this._resolveDescriptor(target, name, _descriptor);
         const { value, writable, enumerable, configurable, get, set } = descriptor;
-        const isFirstInstance = !InstanceChainMap.has([ target, name ]);
-        const chainData = InstanceChainMap.get([ target, name ]) || { fns: [], properties: [] };
+        const isFirstInstance = !InstanceChainMap.has([target, name]);
```
Comment (steelsojka) [inline src/factory/DecoratorFactory.ts]: Auto formatting?

File: package.json (hunk 1 above)
Comment (alan-agius4) [reply package.json]: peers dont get installed automatically

File: src/factory/DecoratorFactory.ts (hunk 3 above)
Comment (alan-agius4) [reply src/factory/DecoratorFactory.ts]: Yeo

File: package.json (hunk 1 above)
Comment (steelsojka) [reply package.json]: That is intentional. It should use your applications version of lodash. This avoids having lodash version mismatches.

File: src/factory/DecoratorFactory.ts (hunk 3 above)
Comment (steelsojka) [reply src/factory/DecoratorFactory.ts]: Can you rollback any formatting that was autoformatted to reduce the scope of changes?

File: package.json (hunk 1 above)
Comment (steelsojka) [reply package.json]: How is this going to resolve if my application is using lodash v4.3? Is it going to use v4.3 for the decorators or is it going to install version 4.17.4 along with it?

File: package.json [hunk 4]
```diff
@@ -58,18 +60,19 @@
...
     "ts-node": "^3.0.2",
+    "tslib": "^1.7.1",
     "tslint": "^5.1.0",
     "tslint-language-service": "^0.9.2",
     "typescript": "~2.3.4"
   },
   "peerDependencies": {
-    "lodash": "4.x"
+    "lodash": "4.x",
+    "tslib": "^1.7.1"
```
Comment (steelsojka) [inline package.json]: tslib should be a dependency. The user shouldn't be required to install a library dependent on how this library is built. Lodash should be a peer and tslib should be a dependency.

File: src/after.ts (hunk 2 above)
Comment (alan-agius4) [reply src/after.ts:1]: `require` won't work when the module is es2015

Comment (steelsojka) [reply src/after.ts:1]: The problem is this will import the entire library for anybody who is not using tree shaking.

**Comment by alan-agius4** [comment]
Well, i do get your point, and I do agree for backward compatability, but
should this be a disadvantage on the library and users using modern and
up-to date tools?

It's your call :-)

**Comment by steelsojka** [comment]
@alan-agius4 Correct me if I'm wrong, but won't importing the methods directly be the same as tree shaking? Importing a single function should only bundle the function any any other files it references. How is tree shaking going to be different?

**Comment by alan-agius4** [comment]
@steelsojka indeed `tree-shaking`  is when the you only bundle the reference function, class etc... The difference is with es2015 you don't need to configure anything for tree-shaking happens outside of the box, and sometimes umd tree-shaking is not even possible in some cases with some bundlers.

**Rollup**
for example, you'd need to use a separate plugin `rollup-plugin-commonjs`, as by default rollup only supports es6 module imports and exports.

I made a small demo showing that umd cannot be used without a plugin (which is still fairly popular to use because not all libraries are using es6 imports) , https://rollupjs.org/repl?version=0.45.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMHVzZWRFczIwMTUlMjAlN0QlMjBmcm9tJTIwJTVDJTIyLiUyRmVzMjAxNSU1QyUyMiUzQiU1Q25pbXBvcnQlMjAlN0IlMjB1c2VkVW1kJTIwJTdEJTIwZnJvbSUyMCU1QyUyMi4lMkZ1bWQlNUMlMjIlM0IlNUNuJTVDbiUyRiUyRiUyMHVzZWRFczIwMTUoKSUzQiU1Q24lMkYlMkYlMjB1c2VkVW1kKCklM0IlMjAlMkYlMkYlMjB1bmNvbW1lbnQlMjBhcyUyMHdpdGhvdXQlMjBhJTIwcGx1Z2luJTIwdW1kJTIwY2Fubm90JTIwYmUlMjB1c2VkJTIwYW5kJTIweW91J2xsJTIwZ2V0JTIwYW4lMjBlcnJvciUyMiU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJlczIwMTUuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwZnVuY3Rpb24lMjB1c2VkRXMyMDE1KCklMjAlN0IlNUNuJTVDdGNvbnNvbGUubG9nKCU1QyUyMnVzZWRFczIwMTUlNUMlMjIpJTNCJTVDbiU3RCU1Q24lNUNuZXhwb3J0JTIwZnVuY3Rpb24lMjBVblVzZWRFczIwMTUoKSUyMCU3QiU1Q24lNUN0Y29uc29sZS5sb2coJTVDJTIyVW5Vc2VkRXMyMDE1JTVDJTIyKSUzQiU1Q24lN0QlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIydW1kLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwdXNlZFVtZCgpJTIwJTdCJTVDbiU1Q3Rjb25zb2xlLmxvZyglNUMlMjJ1c2VkVW1kJTVDJTIyKSUzQiU1Q24lN0QlNUNuJTVDbmZ1bmN0aW9uJTIwVW5Vc2VkVW1kKCklMjAlN0IlNUNuJTVDdGNvbnNvbGUubG9nKCU1QyUyMlVuVXNlZFVtZCU1QyUyMiklM0IlNUNuJTdEJTVDbiU1Q25leHBvcnRzLnVzZWRVbWQlMjAlM0QlMjB1c2VkVW1kJTNCJTVDbmV4cG9ydHMuVW5Vc2VkVW1kJTIwJTNEJTIwVW5Vc2VkVW1kJTNCJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybW9kdWxlTmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMm1vZHVsZUlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=


**Webpack**
Webpack for example, cannot remove unused imports unless it's es6 imports (Maybe there is a plugin but I didn't really look for one) If you see the output code, (the input code is the same as the above) we'll have the below;

```
webpackJsonp([1], {

/***/ "./src/es2015.js":
/***/ (function (module, __webpack_exports__, __webpack_require__) {

		"use strict";
/* harmony export (immutable) */ __webpack_exports__["a"] = usedEs2015;
		/* unused harmony export UnUsedEs2015 */
		function usedEs2015() {
			console.log("usedEs2015");
		}

		function UnUsedEs2015() {
			console.log("UnUsedEs2015");
		}

		/***/
}),

/***/ "./src/main.js":
/***/ (function (module, __webpack_exports__, __webpack_require__) {

		"use strict";
		Object.defineProperty(__webpack_exports__, "__esModule", { value: true });
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_0__es2015__ = __webpack_require__("./src/es2015.js");
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_1__umd__ = __webpack_require__("./src/umd.js");
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_1__umd___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_1__umd__);



		Object(__WEBPACK_IMPORTED_MODULE_0__es2015__["a" /* usedEs2015 */])();
		Object(__WEBPACK_IMPORTED_MODULE_1__umd__["usedUmd"])();
}),

/***/ "./src/umd.js":
/***/ (function (module, exports) {

		function usedUmd() {
			console.log("usedUmd");
		}

		function UnUsedUmd() {
			console.log("UnUsedUmd");
		}

		exports.usedUmd = usedUmd;
		exports.UnUsedUmd = UnUsedUmd;

	})

});

```

Note that `unused harmony export` is only present for `UnUsedEs2015` the code relevant to this will be stripped when using `uglifyjs`

final output
```
webpackJsonp([1], { "./src/es2015.js": function (s, e) { "use strict"; function n() { console.log("usedEs2015") } e.a = n }, "./src/main.js": function (s, e, n) { "use strict"; Object.defineProperty(e, "__esModule", { value: !0 }); { var c = n("./src/es2015.js"), o = n("./src/umd.js"); n.n(o) } Object(c.a)(), Object(o.usedUmd)() }, "./src/umd.js": function (s, e) { function n() { console.log("usedUmd") } function c() { console.log("UnUsedUmd") } e.usedUmd = n, e.UnUsedUmd = c } });
```

**Comment by steelsojka** [comment]
@alan-agius4 I understand what you are saying but we are ending up with very similar results due to the modularized way lodash is broken up. If we did the es2015 imports, bundling tools could do it for us but requires the user to configure their tool to perform the tree shaking, when we are already importing just the required code by importing the functions needed. Ideally we could use `lodash-es` to do import statements like `import get from 'lodash-es/get';` but then we can't use the users lodash cause they will most likely have `lodash` installed and not `lodash-es`.

**Comment by alan-agius4** [comment]
I get your point as well, i'll try to figure out something maybe over the weekend 😊 But i agree that i prefer to use lodash over lodash-es and if we change that we won't be able to use the users lodash version

**Comment by alan-agius4** [comment]
I'll close this for now, as really I cannot find a way to import a method only from lodash

**Pull request closed without merging**